> macOS failed

test.pythoninfo failed with:

```
ERROR: collect_pwd() failed
Traceback (most recent call last):
  File "/Users/runner/runners/2.165.2/work/cpython/cpython/Lib/test/pythoninfo.py", line 761, in collect_info
    collect_func(info_add)
  File "/Users/runner/runners/2.165.2/work/cpython/cpython/Lib/test/pythoninfo.py", line 336, in collect_pwd
    groups = os.getgrouplist(entry.pw_name, entry.pw_gid)
OSError: [Errno 0] Error
```

test_ttk_guionly failed on Ubuntu:


```
(...)
test_initialization_no_master (tkinter.test.test_ttk.test_extensions.LabeledScaleTest) ... ok
test_resize (tkinter.test.test_ttk.test_extensions.LabeledScaleTest) ... ok
test_variable_change (tkinter.test.test_ttk.test_extensions.LabeledScaleTest) ... Timeout (0:20:00)!
Thread 0x00007fa8861ec080 (most recent call first):
  File "/home/runner/work/cpython/cpython/Lib/tkinter/__init__.py", line 697 in wait_visibility
  File "/home/runner/work/cpython/cpython/Lib/tkinter/test/test_ttk/test_extensions.py", line 147 in test_variable_change
  File "/home/runner/work/cpython/cpython/Lib/unittest/case.py", line 616 in _callTestMethod
(...)
```

These two errors are unrelated to my PR.

--- a/Doc/whatsnew/3.9.rst
+++ b/Doc/whatsnew/3.9.rst
@@ -411,6 +411,16 @@ Optimizations
 
   (Contributed by Serhiy Storchaka in :issue:`32856`.)
 
+* Optimize signal handling in multithreaded applications. If a thread different
+  than the main thread gets a signal, the bytecode evaluation loop is no longer
+  interrupted at each bytecode instruction to check for pending signals which
+  cannot be handled. Only the main thread of the main interpreter can handle
+  signals.
+
+  Previously, the bytecode evaluation loop was interrupted at each instruction
+  until the main thread handles signals.
+  (Contributed by Victor Stinner in :issue:`40010`.)
+
 
 Build and C API Changes
 =======================

--- a/Misc/NEWS.d/next/Core and Builtins/2020-03-19-02-26-13.bpo-40010.Y-LIR0.rst
+++ b/Misc/NEWS.d/next/Core and Builtins/2020-03-19-02-26-13.bpo-40010.Y-LIR0.rst
@@ -1,0 +1,8 @@
+Optimize signal handling in multithreaded applications. If a thread different
+than the main thread gets a signal, the bytecode evaluation loop is no longer
+interrupted at each bytecode instruction to check for pending signals which
+cannot be handled. Only the main thread of the main interpreter can handle
+signals.
+
+Previously, the bytecode evaluation loop was interrupted at each instruction
+until the main thread handles signals.

--- a/Python/ceval.c
+++ b/Python/ceval.c
@@ -120,6 +120,15 @@ static size_t opcache_global_hits = 0;
 static size_t opcache_global_misses = 0;
 #endif
 
+
+/* Only handle signals on the main thread of the main interpreter. */
+static int
+thread_can_handle_signals(void)
+{
+    return (PyThread_get_thread_ident() == _PyRuntime.main_thread);
+}
+
+
 /* This can set eval_breaker to 0 even though gil_drop_request became
    1.  We believe this is all right because the eval loop will release
    the GIL eventually anyway. */
@@ -127,7 +136,8 @@ static size_t opcache_global_misses = 0;
     _Py_atomic_store_relaxed( \
         &(ceval2)->eval_breaker, \
         _Py_atomic_load_relaxed(&(ceval)->gil_drop_request) | \
-        _Py_atomic_load_relaxed(&(ceval)->signals_pending) | \
+        (_Py_atomic_load_relaxed(&(ceval)->signals_pending) \
+            && thread_can_handle_signals()) | \
         _Py_atomic_load_relaxed(&(ceval2)->pending.calls_to_do) | \
         (ceval2)->pending.async_exc)
 
@@ -156,10 +166,11 @@ static size_t opcache_global_misses = 0;
         COMPUTE_EVAL_BREAKER(ceval, ceval2); \
     } while (0)
 
+/* eval_breaker is not set to 1 if thread_can_handle_signals() is false. */
 #define SIGNAL_PENDING_SIGNALS(ceval, ceval2) \
     do { \
         _Py_atomic_store_relaxed(&(ceval)->signals_pending, 1); \
-        _Py_atomic_store_relaxed(&(ceval2)->eval_breaker, 1); \
+        COMPUTE_EVAL_BREAKER(ceval, ceval2); \
     } while (0)
 
 #define UNSIGNAL_PENDING_SIGNALS(ceval, ceval2) \
@@ -540,8 +551,7 @@ handle_signals(PyThreadState *tstate)
 {
     _PyRuntimeState *runtime = tstate->interp->runtime;
 
-    /* Only handle signals on main thread */
-    if (PyThread_get_thread_ident() != runtime->main_thread) {
+    if (!thread_can_handle_signals()) {
         return 0;
     }
     /*

--- a/Python/ceval_gil.h
+++ b/Python/ceval_gil.h
@@ -280,8 +280,17 @@ _ready:
     COND_SIGNAL(gil->switch_cond);
     MUTEX_UNLOCK(gil->switch_mutex);
 #endif
+
     if (_Py_atomic_load_relaxed(&ceval->gil_drop_request)) {
         RESET_GIL_DROP_REQUEST(ceval, ceval2);
+    }
+    else {
+        /* bpo-40010: eval_breaker should be recomputed to be set to 1 if there
+           a pending signal: signal received by another thread which cannot
+           handle signals.
+
+           Note: RESET_GIL_DROP_REQUEST() calls COMPUTE_EVAL_BREAKER(). */
+        COMPUTE_EVAL_BREAKER(ceval, ceval2);
     }
 
     int must_exit = tstate_must_exit(tstate);

